### ISSUE TITLE
Let multiple containers share downloaded dependencies

### ISSUE THREAD

**Issue by bsideup** [gradle/gradle#851]
Hi!

Looks like Gradle is locking the global cache when running the tests. We run Gradle in Docker containers, and from what I saw in the logs, it fails to acquire the lock with:
```
14:09:05.981 [DEBUG] [org.gradle.cache.internal.DefaultFileLockManager] The file lock is held by a different Gradle process (pid: 1, operation: ). Will attempt to ping owner at port 39422
``` 

### Expected Behavior
Gradle should release the lock when it executes the tests. Other Gradle instances are failing with:
```
Timeout waiting to lock Plugin Resolution Cache (/root/.gradle/caches/3.2/plugin-resolution). It is currently in use by another Gradle instance.
```

### Current Behavior
Gradle holds the lock

### Context
Our CI servers are affected. Parallel builds are impossible.

### Steps to Reproduce
I managed to reproduce it with a simple Docker-based environment:
https://github.com/bsideup/gradle-lock-bug

 Docker and Docker Compose should be installed.
```
$ git clone git@github.com:bsideup/gradle-lock-bug.git
$ COMPOSE_HTTP_TIMEOUT=7200 docker-compose up --no-recreate
```

### Your Environment
Gradle 3.2 (tried with 3.1, 3.0 and 2.12 as well)
Docker

**Comment by oehme** [comment]
I don't quite understand the use case yet. Are you running several builds at the same time on the same working directory? That will give you many other odd problems besides just the .gradle directory being locked. Just think about what happens when one of those builds runs a `clean` while another is trying to compile.

If you want to do different builds on the same project at the same time, I'd recommend using separate checkouts for that.

**Comment by bsideup** [comment]
@oehme 
We don't, I just reused the same project source to demonstrate an issue. We run different projects inside the containers at the same time with .gradle folder shared across them (think CI environment)

**Comment by oehme** [comment]
Got it, thanks. The Gradle user home cannot be shared between different machines. Why do you want to share it? It'll just create contention between your builds, even if this specific issue was solved.

**Comment by bsideup** [comment]
@oehme well, it's a bit hard to define "machine" here. 
We use Docker containers, at the same host. 

Having to create a separate .gradle **for each project** sounds a bit expensive and breaks the concept of the shared global cache.

**Comment by oehme** [comment]
A docker container is a machine for that matter. It's processes are isolated from the host system.

> Having to create a separate .gradle for each project sounds a bit expensive and breaks the concept of the shared global cache.

I don't really understand the use case I guess. What is the reason to run the builds in docker containers, but share the user home? If you don't trust the code, then it absolutely should not have write access to the host's user home. If you trust it, then what do the docker containers buy you?

**Comment by bsideup** [comment]
@oehme it's not about the security. 

We use Docker containers as a unified way to run different kinds of builds in our CI process, different projects might want to use different Java versions, for instance

This is pretty common things nowadays I should say. Jenkins is promoting Docker builds a lot, others are integrating Docker containers as well. 

I understand the problem with "the multiple machines issue". However, this issue is more about "Why test executor takes a lock for a long time?", because AFAIK locks in Gradle are short living things.

**Comment by oehme** [comment]
Gradle processes will hold locks if they are uncontended (to gain performance). Contention is announced through inter-process communication, which does not work when the processes are isolated in Docker containers.

**Comment by bsideup** [comment]
Hm, back in the days it was different - Gradle was trying to release the lock as soon as possible, and I really liked that strategy. What happened? :)

**Comment by oehme** [comment]
The cross-process caches use file-based locking, so every lock/unlock operation is an I/O operation. Since these are expensive, we try to avoid them as much as possible.

**Comment by bsideup** [comment]
Any chance to get them configurable? I would really like to disable this optimization on our CI environments. Otherwise, we just delete lock file manually to workaround the issue when there are some long-running tests are being executed :D

**Comment by oehme** [comment]
We could potentially add a system property that tells Gradle to "assume contention". There might be other issues that we haven't yet discovered though, since sharing a user home between machines is not a use case we have designed for.

I'd like to assess the alternatives first: What would be the drawback if you don't share the user home?

**Comment by bsideup** [comment]
A **huge** disk space and network usage. We will have to download the same dependencies for every Gradle job type. Right now Gradle cache takes a few GBs, but if we don't share, we will have to multiply it by the number of Gradle-based jobs we have, so the result will be tens, or maybe even hundreds of GBs, which is not really acceptable for us

**Comment by oehme** [comment]
I think the best next step would be for you to [implement a fix](https://github.com/gradle/gradle/blob/afaa23af3a86ea41c3dfe91bf26973bcb719f818/subprojects/core/src/main/java/org/gradle/cache/internal/DefaultCacheAccess.java#L419) for that specific problem and try it out in your environment. 

My gut feeling is that there may be other issues waiting when you try to reuse the user home. If there aren't, then we could discuss introducing a flag into Gradle to opt-in to a "docker mode" :)

**Comment by bsideup** [comment]
@oehme ok, thanks for the link! I'll try to play around with it and will report back.

Also, there is one more option - on *nix-based systems, Gradle can use sockets to communicate. That way it should work, and Docker will allow us to mount the socket inside a container.

WDYT?

**Comment by oehme** [comment]
That could work ad well. Let's first make sure though that the locking problem is in fact the only problem here.

**Comment by saiimons** [comment]
@bsideup Did you fix this ? I am currently facing this issue with the same kind of setup as yours...
At least, it would be nice to have an option to set the timeout.

**Comment by martinda** [comment]
Another use case is when a user runs multiple different builds of different projects on multiple different hosts, all using his/her account. This is typical of environments with network mounted home directories.

Gradle has to pro-actively release the lock as soon as it is done with the cache. I am willing to pay the price of an IO operation to save the build from a timeout. Please see the excellent explanation in this [GRADLE-3106](https://issues.gradle.org/browse/GRADLE-3106?focusedCommentId=21108&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-21108) comment.

**Comment by martinda** [comment]
Just want to explain how to reproduce this problem by posting a simple `build.gradle` file:
```
task sleep() {
    doLast {
        Thread.sleep(100000)
    }
}
````
Get two terminals on different hosts that mount the same home directory with the same `~/.gradle` in it, then type `gradle sleep --debug --stacktrace` in both terminals. One of them will fail to acquire the lock and die waiting. The failing one will show:
```
The file lock is held by a different Gradle process (pid: 64549, operation: ). Will attempt to ping owner at port 40291
```
Of course the other process cannot be notified, it is on another host, resulting in:
```
Caused by: org.gradle.cache.internal.LockTimeoutException: Timeout waiting to lock file hash cache (/home/martinda/.gradle/caches/3.5/fileHashes). It is currently in use by another Gradle instance.
    Owner PID: 64549
    Our PID: 25504
    Owner Operation: 
    Our operation: 
    Lock file: /home/martinda/.gradle/caches/3.5/fileHashes/fileHashes.lock
```
Could it be as simple as adding the IP address of the process holding the lock to the lock file and add it to the [pingOwner](https://github.com/gradle/gradle/blob/v3.5.0/subprojects/core/src/main/java/org/gradle/cache/internal/FileLockCommunicator.java#L44) method?

**Comment by AdrianAbraham** [comment]
My team is also encountering this issue when dealing with containerized CI builds, forcing us to keep many copies of the Gradle cache. We'd love to see an option to aggressively release the lock.

**Comment by bsideup** [comment]
FYI: 
For us, the workaround was to run the container where Gradle is with `--net=host`, this way Gradle will be able to communicate with other instances.

**Comment by saiimons** [comment]
My workaround is setting up a maven repository acting as a proxy (and cache) and loading some init.gradle in the build container.
This allows us to keep one cache instead of multiple ones, and there is no conflict.

**Comment by AdrianAbraham** [comment]
@bsideup That sounds good, though I don't know if I'll be able to convince my CI runner to do that ... I'll try it out.

@saiimons Could you go into more detail about your workaround?

**Comment by saiimons** [comment]
@AdrianAbraham I run a Nexus repository with a proxy configuration for the major maven repositories (jcenter, maven central, etc. check your log for the URLs).

![image](https://cloud.githubusercontent.com/assets/6832416/25954124/539d25ec-3633-11e7-8528-0ef958431caa.png)

All these guys go behind a group, in order to use a single URL:

![image](https://cloud.githubusercontent.com/assets/6832416/25954200/88c6c408-3633-11e7-8d0b-cd895f30b03b.png)

Then my build container will load a init.gradle file (in /opt/gradle/init.d/ as I am using [this image](https://hub.docker.com/r/frekele/gradle/) for CI) :
```groovy
allprojects {
  buildscript {
    repositories {
      mavenLocal()
      maven {
        url "http://nexus:8081/repository/global_proxy/"
      }
    }
  }
  repositories {
    mavenLocal()
    maven {
      url "http://nexus:8081/repository/global_proxy/"
    }
  }
}
```

**Comment by AdrianAbraham** [comment]
@saiimons We're running a local Nexus, and our configuration is similar (no mavenCentral(), though); but Gradle still has to download the packages from Nexus into its own cache to run a build. Does your setup just avoid Internet downloads? Or does it avoid the Gradle cache itself?

**Comment by saiimons** [comment]
Yes, gradle downloads the packages, but as the storage is local, the latency is small and there is no bandwidth consumed.
My build sped up and the need for external resources was reduced (we were able to build when DNS servers were DDoSed in October and most of the repositories were unreachable).

**Comment by hexsel** [comment]
We're doing something similar, but using automatically-generated docker containers for each branch (this is handled by Jenkins 2.0 pipelines and it's _somewhat_ flexible).  We still have to download 2GB or so worth of jars for every build of every branch, and keep storage for all of these until Jenkins auto-cleans them a few days later. 

It's not a deal breaker, but it's a serious inconvenience.  I could live with an occasional contention on the folders if it made the downloads unnecessary.

Is there any way to configure the timeout for the cache?

**Comment by andrask** [comment]
We used to solve this issue by generating a base image that cached most of the dependencies. This base image was rebuilt every day, thus, the actual difference between the current and cached dependencies was kept small.

In our new project, however, we are facing the same issue as @bsideup as we want to share the cache between containers (the above scheme works for this). At the same time we are limited by the 'concurrent builds cannot use the same cache' aspect. Actually, the real issue is not even the cache size or the network usage, but the serialized download of artifacts. We have 10G connection between the servers, (I even put the repos on RAM disk,) still the network utilization is low as the dependencies are downloaded sequentially. Probably even over different TCP connections, resulting in lots of time wasted for TCP ramp up. This may be a result of using Maven repos. The real solution would probably be a server side resolver framework that could return all the artifact URLs at once.

**Comment by andrask** [comment]
We used to solve this issue by generating a base image that cached most of the dependencies. This base image was rebuilt every day, thus, the actual difference between the current and cached dependencies was kept small.

In our new project, however, we are facing the same issue as @bsideup as we want to share the cache between containers (the above scheme works for this). At the same time we are limited by the 'concurrent builds cannot use the same cache' aspect. Actually, the real issue is not even the cache size or the network usage, but the serialized download of artifacts. We have 10G connection between the servers, (I even put the repos on RAM disk,) still the network utilization is low as the dependencies are downloaded sequentially. Probably even over different TCP connections, resulting in lots of time wasted for TCP ramp up? This may be a result of using Maven repos. The real solution would probably be a server side resolver framework that could return all the artifact URLs at once.

**Comment by oehme** [comment]
> Actually, the real issue is not even the cache size or the network usage, but the serialized download of artifacts.

Gradle 4.0-milestone-2 downloads metadata and artifacts in parallel, you might wanna give that a try.

**Comment by andrask** [comment]
> Gradle 4.0-milestone-2 downloads metadata and artifacts in parallel, you might wanna give that a try.

Definitely very promising. Once 4.0 comes out, I'll push for the upgrade.

**Comment by zageyiff** [comment]
Hello,
We have similar setup, docker containers as jenkins slaves. 
We're trying to have the gradle user home as a mounted volume on the docker container so all containers reuse the gradle cache and avoids downloading the common 3pp dependencies.

Using Gradle 4 RC 3, still does not fix the issue
FAILURE: Build failed with an exception.

* What went wrong:
Could not create service of type FileHasher using GradleUserHomeScopeServices.createCachingFileHasher().
> Timeout waiting to lock file hash cache (/srv/jenkins/workspace/.gradle/caches/4.0-rc-3/fileHashes). It is currently in use by another Gradle instance.
  Owner PID: 170
  Our PID: 170
  Owner Operation: 
  Our operation: 
  Lock file: /srv/jenkins/workspace/.gradle/caches/4.0-rc-3/fileHashes/fileHashes.lock

As of now, we need to set gradle user home unique per container so each build is downloading all dependencies.
We're not able to set --net-host, as we have some services in the container that are used as part of integration test (postgres) and we would need to have different port per container

**Comment by mkobit** [comment]
We see this happen periodically in our Jenkins build fleet as well.

The general use pattern is:

* 1 executor per node (so there are not multiple builds executing in parallel or multiple Gradle builds running sharing the same home directory)
* Home directory is (currently) shared across multiple from multiple different repositories and Gradle versions
* The build is ran inside of a container
* User may run many shell commands or `./gradlew` commands inside of that container
* The container is provisioned for as long as the user decides
* A user may use the daemon inside the container as the may run various different
* Container is stopped at the end of the user's build so any Gradle daemon running inside that container is also stopped

We then sometimes see the same error mentioned by @zageyiff for users' builds that execute on that node.

> Could not create service of type FileHasher using GradleUserHomeScopeServices.createCachingFileHasher()

What we see happens in some builds execute successfully on a node. They may fail or be manually killed. Then, a Gradle execution will finish with the error above.

I don't have an easy way to reproduce it yet, but if there is anything useful I can provide here let me know.

**Comment by gayakwad** [comment]
Is it possible to point gradle to an additional read-only local cache ? If this can be achieved a read only volume can be mounted on a container, which can avoid multiple user and permission related problem as well.

**Comment by RMsiemens** [comment]
We observe this on our docker based Gitlab runners as well, as we also have one shared cache for all runners. We share the cache by mounting the cache folder into every docker container.

We have now a workaround in place. We do not any longer share the cache but have individual caches in each docker container. This results in slower build times and uses a lot of resources (Storage, bandwith, ...).

Would be nice if the proposed new gradle system property "assume contention" could be implemented.

**Comment by dforrest88** [comment]
This issue is causing a lot of pain for me as well. We're trying to not rely on build node dependencies and thus running all of our gradle tasks inside of docker containers. Given that we also have several gradle utilities that we use to execute fine-grained build and environment tasks with a high degree of parallelism, each task takes significantly longer than it should because it has to fetch all of the dependencies. It would be nice to be able to mount gradle home on the build node each time leverage the shared cache similarly to how it can be done in maven.

Does anyone have an example for how they implemented pre-cached containers? Do you link a data container on gradle home that is pre-populated with the expected cache?

**Comment by aaroncline** [comment]
Similar to @RMsiemens, I'm looking at using a multi-stage Docker build on Gitlab Runners to build Java apps and a subsequent application container.  I also plan to use Gitlab Runners to run CI tests on commits and merged in Docker containers.  I want to do this to more easily offer multiple Java versions to our dev teams.   

I would appreciate a gradle cache sharing solution allowing me to have a common cache on the Gitlab Runners that I can share to multiple Docker containers running builds or tests that may be happening at the same time.  We use a remote artifact store and will have to pay for massive amounts of bandwidth to perform the download of dependencies each time a new build or merge happens and all of the dependencies are downloaded.

**Comment by mdekstrand** [comment]
Another use case for this is using Gradle for automation in an HPC environment, particularly with different tasks that the user may want to run in parallel in a computing cluster.

These clusters typically have high-performance networked file systems.

Right now, wrangling Gradle to work on HPC systems is not easy.

**Comment by costincaraivan** [comment]
I hope this can be reprioritized. Building in Docker containers is becoming more and more common, especially for CI, and since containers usually start from a "clean" state, this means that each and every build triggers an artifact download. The alternative is to bake the Gradle cache in the container, which is really hacky and also makes the containers really big.

The cleanest solution, by far, is Gradle being able to share the same cache without any conflicts between multiple instances. I appreciate Gradle, but c'mon folks, (conceptually) it shouldn't be that hard, Maven 2 was doing it without breaking a sweat back in 2008 😢

**Comment by Rushit** [comment]
We are facing exact same problem. Please let us know when should we expect fix for this.

**Comment by martino-letgo** [comment]
Hello guys,
I am facing the same issues while trying to run multiple Dockers that share the same cache. Any workaround?

**Comment by sschuberth** [comment]
A possible work-around that was communicated to us by Gradle support is:

> [...] by running a remote cache container per docker host and arranging for the containers doing builds with the build tool to use that endpoint for remote caching. Gradle Enterprise also has a high performance shared remote cache backend that you could try out for free.

And:

> We make a docker container available of a remote cache: https://hub.docker.com/r/gradle/build-cache-node/

**Comment by Tapchicoma** [comment]
But it sounds more like caching the tasks output and not for caching the dependencies.

Also possible workaround would be spinning up proxy artifcatory image in same as CI network.

**Comment by anderslauri** [comment]
To workaround this issue we let our containers (jenkins nodes) rsync the entire gradle cache to a common persistent volume when they are destroyed after each unique build process and during startup of the containers the cache is reloaded from the persistent volume using rsync as well to in memory on the jenkins node - it's a simple solution using merely shell scripting but it's effective for us and rsync deals with alot of the complexity if files exists etc, in combination we do use the gradle build cache and nexus as well locally within our CI/CD-cluster but those changes only made minor performance improvements.

However it would be nice not to be locking files when reading from the cache...

**Comment by FrailWords** [comment]
Another workaround and not a good one is to set different local cache folders for each Gradle build - http://mrhaki.blogspot.in/2017/04/gradle-goodness-change-local-build.html.  This is not optimal but is simple.

**Comment by dzwicker** [comment]
Daniel Zwicker
Schulstraße 45b
64342 Seeheim-Jugenheim

**Comment by Moritz90** [comment]
FWIW, I'm getting locking errors even though I'm using different GRADLE_USER_HOMEs for each job. Builds will sporadically fail with the message:

> Could not create service of type FileHasher using GradleUserHomeScopeServices.createCachingFileHasher()

All GRADLE_USER_HOMEs are on NFS.

It seems the file locking issues in Gradle are not limited to concurrent use of the same GRADLE_USER_HOME.

**Comment by marmax** [comment]
@anderslauri would you mind to share the scripts (jenkinsfile, bash, whatever), if possible?

**Comment by anderslauri** [comment]
@marmax 

It's simple, stupid simple. Below is the snippet for our OpenShift configuration for our Jenkins nodes (OpenShift pod) which executes the following script once it's get killed after each unique Jenkins build.

```
  - image: nexus.kafkaint.fhm.de/stargate/jenkins-node:latest
    lifecycle:
      preStop:
        exec:
          command: ["/home/jenkins/bin/stop"]
```

The script stop contains the following where the variable GRADLE_CACHE is the directory for the Gradle build cache in memory on the pod and the respective NFS variables are PVC mounted directories on the pod.

```
if [ -d "${GRADLE_CACHE}" -a -d "${GRADLE_CACHE_NFS}" ]; then
  rsync --whole-file --ignore-existing --recursive "${GRADLE_CACHE}/" "${GRADLE_CACHE_NFS}"
fi

if [ -d "${SONAR_CACHE}" -a -d "${SONAR_CACHE_NFS}" ]; then
  rsync --whole-file --ignore-existing --recursive "${SONAR_CACHE}/" "${SONAR_CACHE_NFS}"
fi
```

When a new pod is started the entrypoint script just contains this below. All of this works very fine for us, we use a cronjob in OpenShift to clean the cache on the NFS once a month to ensure a relevant cache also.

```
  # Synchronize the Gradle cache to the container.
  if [ -d "${GRADLE_CACHE_NFS}" -a -d "${GRADLE_CACHE}" ]; then
    nohup rsync --whole-file --ignore-existing --recursive "${GRADLE_CACHE_NFS}/" "${GRADLE_CACHE}" > /dev/null 2>&1 &
  fi
  
  # Synchronize the SonarQube cache to the container.
  if [ -d "${SONAR_CACHE_NFS}" -a -d "${SONAR_CACHE}" ]; then
    nohup rsync --whole-file --ignore-existing --recursive "${SONAR_CACHE_NFS}/" "${SONAR_CACHE}" > /dev/null 2>&1 &
fi
```

**Comment by ntkoopman** [comment]
Is there any stance from the Gradle team about this? Is this something that should be supported or are we supposed to go all in on the build cache and a local repository? All I see here are workarounds.

**Comment by oehme** [comment]
The answer depends on the problem you are trying to solve. Each of them might have different solutions.

1. Dependencies being re-downloaded? Use Nexus/Artifactory close to your build agents (e.g. as a container on the same machine). We might also put downloaded dependencies into the Gradle build cache to serve a similar purpose.
2. Too much disk space being used because each agent stores its own copy of the dependencies? If you use ephemeral agents that are cleaned up after each build this should not be an issue. If you use long-lived agents we could work against the cache growth by implementing something for #1085
3. You want to share configuration like gradle.properties or init scripts? Only copy those into the agents. We might separate a configuration dir and a cache dir in the future.

Any other use cases I'm missing?

**Comment by bsideup** [comment]
@oehme yes. The reported one :D

4. Concurrent access to `~/.gradle` folder from different Gradle daemons / instances. 

Currently, they use `localhost` to communicate with each other. If you change it to file socket, for instance, it shouldn't be a problem anymore.

**Current workaround for Docker users:**
Run your containers with `--net=host`, so that Gradle instances will communicate with each other

**Comment by gesellix** [comment]
Another possible workaround (I haven't tried, yet): put all containers in the same network.

**Comment by bsideup** [comment]
@gesellix AFAIR that will not work because Gradle will still use `localhost` :(

**Comment by gesellix** [comment]
@bsideup ah, you're right, literally `localhost`.

**Comment by oehme** [comment]
@bsideup That's not a use case, but an implementation detail. You are doing this because you want to solve one or multiple of the use cases I listed. 

Allowing concurrent access to `~/.gradle` would have weird side effects. E.g. suddenly one container would try to run builds on a daemon from another container, because they both see the same daemon registry. Also, the caches in there are OS-dependent, so if you have VMs with both Linux and Windows for instance, that wouldn't work either. 

So for the limited use case of "Docker containers on the same machine" we could make this work by separating the caches from other state and making that sharable. But for all other use cases this wouldn't help. Plus, it might mean losing some performance for the common case of a non-shared home. 

I think a better/more general solution would be to put dependencies in the build cache and then have a build cache node on the same physical machine as the build agents.

**Comment by costincaraivan** [comment]
By node you mean an actual container? I guess that could work, even though it would make things more complex... What would creating a "build cache node" involve?

**Comment by oehme** [comment]
@costincaraivan See https://hub.docker.com/r/gradle/build-cache-node/

**Comment by bsideup** [comment]
> I think a better/more general solution would be to put dependencies in the build cache and then have a build cache node on the same physical machine as the build agents.

You keep repeating that, or "start local Nexus for Maven repository', but... Really? That's a hack/workaround, not a solution. It's even worse when multiple teams/companies share the same CI infrastructure like Jenkins.

> Allowing concurrent access to ~/.gradle would have weird side effects. E.g. suddenly one container would try to run builds on a daemon from another container, because they both see the same daemon registry.

That's not a side effect but bug if you ask me.

> So for the limited use case of "Docker containers on the same machine" we could make this work by separating the caches from other state and making that sharable

Sounds like a great separation of concerns which is broken right now. IMO it should be fixed not because of #851 (this issue), but also because of any other potential issue it might cause.

**Comment by Moritz90** [comment]
I agree with @bsideup: The cache should be shareable between containers/machines with access to the same file system. It's simply what users expect. And cleanly separating the caches from all other state sounds like a very good idea in general.

**Comment by martinda** [comment]
> We might separate a configuration dir and a cache dir in the future.

There is an issue for this already: https://github.com/gradle/gradle/issues/1319

My users login to multiple hosts. These hosts are shared between multiple users. We don't mix OSes, so it is an homogeneous network of hosts. Any given user can run any number of builds on any number of hosts at the same time. Since all users are pointing to their respective NFS homes, the `~/.gradle` cache is common between those hosts.

If gradle released the cache sooner (after downloading dependencies?) it would help. Maybe this can be a temporary solution?

**Comment by lptr** [comment]
@bsideup suppose we separate caches and allow them to be shared between containers. How would you solve the different OS problem?

**Comment by bsideup** [comment]
@lptr different OS?

**Comment by melix** [comment]
I think we all agree that having a shared cache would be great, for different reasons. However, the fact our cache is _by design_ limited to a single machine, and not meant to be shared over the network or docker containers is not an arbitrary decision to make everybody swear. Ideally, if we can find a solution that:

1. allows concurrent read/write access in the cache
2. from a single machine (multiple builds executed concurrently on a single machine)
3. fine grained (doesn't lock the cache for the lifetime or a build, like it used to, or dependency resolution, like it used to also)
4. supports different OS (including, yes, Windows)
5. supports multiple concurrent hosts (aka, the docker use case here, or different CI agents)
6. doesn't lock the cache for a single host, blocking all others
7. is TCP/IP connection failure safe (in other words, you're not allowed to use TCP/IP to see that there are concurrent access, because docker isolates everybody)
8. doesn't kill the performance of the local builds

Then of course, we would accept such a PR. Today, what we have supports 1 to 4.

**Comment by lptr** [comment]
@bsideup what @oehme mentioned:

> Allowing concurrent access to ~/.gradle would have weird side effects. E.g. suddenly one container would try to run builds on a daemon from another container, because they both see the same daemon registry. **Also, the caches in there are OS-dependent, so if you have VMs with both Linux and Windows for instance, that wouldn't work either.**

**Comment by bsideup** [comment]
@lptr I wonder why does caching of things downloaded from the internet have to be OS dependent? Not talking about Gradle's build cache or something

**Comment by oehme** [comment]
We could reduce the scope of the problem to just the file store (i.e. downloaded POMs and JARs), leaving all other caches private. Since these are written infrequently, there shouldn't be a performance problem when sharing them with a more pessimistic locking strategy. Even with a localized dependency mirror that would still be worthwhile, as the agents would save disk space.

The other caches (e.g. compiled scripts, build cache, transformed artifacts etc.) would be much harder to handle. So we'd need to separate those directories using some new option.

**Comment by bsideup** [comment]
@oehme sounds good to me. The issue was about the files downloaded from the internet (POMs, JARs, etc), everything else is not critical for us

**Comment by gesellix** [comment]
+1 from me for scoping "dependencies" (jars, etc), because we would like to reduce network traffic of bigger files. Regarding the build cache: I assumed that would already be addressed with the global build cache (and disabling local build cache)?

**Comment by ldaley** [comment]
@bsideup just out of curiosity, why are you having to download dependencies over the Internet?  Is it that …

- You do not have an internal proxy (e.g. Artifactory)? 
- You are using a near proxy but it is still too slow?
- You are building on a platform (e.g. Travis) that doesn't support that kind of near proxy? 
- Something else?

@gesellix @gayakwad @mkobit @zageyiff @saiimons - I'd appreciate your answers to if you don't mind. If you prefer, you can email me direct via luke - at - gradle.com. Thanks in advance.

**Comment by bsideup** [comment]
@ldaley,

> You do not have an internal proxy (e.g. Artifactory)?
> You are using a near proxy but it is still too slow?
No, that will only increase the complexity & cost of our build infrastructure

> You are building on a platform (e.g. Travis) that doesn't support that kind of near proxy?
Sometimes, but it's not affected by the issue I described (Although you have to delete some files (some `.lock`s) from the cache, otherwise Gradle will fail... )

> Something else?
A standard setup with Jenkins where every job is executed inside a Docker container (to avoid having to install the tools on a host), multiple containers per host with mounted `~/.gradle`.

**Comment by vyphan** [comment]
@oehme Can you guess what milestone this feature might be targeted for? Just being able to share those POMs and JARs would be a huge speed improvement for teams that build in Docker containers.

**Comment by oehme** [comment]
There is no plan at this point. Our recommendation for now remains using artifactory/nexus to provide fast artifact downloads both for your CI agents and your team members.

**Comment by redeamer** [comment]
In regard of this 'issue' in a CI context:
For our setup I think I introduced a good 'workaround', but it depends on your number of executors per build node and the size of a singe cache: I map a container volume to keep the gradle caches and set the `GRADLE_USER_HOME` to `<cache_volume_path>/${env.EXECUTOR_NUMBER}` (on Jenkins-CI, do not know for other CIs). So I avoid any parallelism issues and have still the cache around for reuse and the cache duplication is justifiable/feasible (for us at least).

**Comment by vyphan** [comment]
@redeamer Have you run into any issues from persisting GRADLE_USER_HOME from build to build? Wondering what else other than the downloaded dependencies is carried over.

**Comment by AdrianAbraham** [comment]
We use almost exactly the same workaround as @redeamer and haven't had any issues

**Comment by michellehalliwell** [comment]
Unfortunately a workaround like the one @redeamer mentioned above won't work with Dockerized slaves.  The build will always be running on executor 1 since each Docker container is treated as a new node.

**Comment by Saggi432** [comment]
@bsideup we are using kubenetes to launch the jenkins agents on containers using the kubenetes plugin. we do face a similar issue when we run multiple containers, which are sharing the same NAS. This approach does help us to fasten our build time..

As per the discussion above, how can we inject the --net=host in the kube environment for the containers as we run. The plugin kubernetes-plugin, does not give us the options to pass the arguments when we run the docker. 


+ ./gradlew clean test --refresh-dependencies
Starting a Gradle Daemon, 1 busy and 1 incompatible and 3 stopped Daemons could not be reused, use --status for details

FAILURE: Build failed with an exception.

* What went wrong:
Could not create service of type FileHasher using GradleUserHomeScopeServices.createCachingFileHasher().
> Timeout waiting to lock file hash cache (/nas/jenkins/.gradle/caches/4.0/fileHashes). It is currently in use by another Gradle instance.
  Owner PID: 141
  Our PID: 141
  Owner Operation: 
  Our operation: 
  Lock file: /nas/jenkins/.gradle/caches/4.0/fileHashes/fileHashes.lock

**Comment by vertrost** [comment]
Our cache takes about ~1.3GB, so the decision was to rsync .gradle for every docker container and than rsync updates back to volume. But still would be nice to have a solution to run docker just from volume, it would be much convenient and will take less time for "syncing".

**Comment by daggerok** [comment]
very actual problem, 

on each docker pipeline build at the moment I'm:

- disabling daemon `org.gradle.daemon=false`

- on non root user [doing chown](https://github.com/daggerok/e2e-ubuntu-jdk8-xvfb/blob/c65a84ea856151993d3c8888f24dffcc63411277/Dockerfile#L36), because all folder are mounting wuth root user

- and mounting no whole ~/.gradle folder, but only dependency caches:

```bash
docker run --rm --name run-my-e2e-tests \
  -v ~/.gradle/caches/modules-2/files-2.1:/home/e2e/.gradle/caches/modules-2/files-2.1 \
  -v ~/.m2/repository:/home/e2e/.m2/repository \
  my-e2e-tests
```

**Comment by richfromm** [comment]
Here is our situation:

We use docker for our CI builds. Not for security; we use docker for repeatability, isolation, and ease of management. Each build runs in its own docker container. We have up to 3 build agents running on the same host, each as a separate docker container.

We recently enabled the gradle build cache, pointing to a remote [HttpBuildCache](https://docs.gradle.org/current/javadoc/org/gradle/caching/http/HttpBuildCache.html). I hadn't realized that the default behavior, if you enable caching, is that you also get a local cache, at `$GRADLE_HOME/caches/build-cache-1`. Note that we have `$GRADLE_HOME` mapped outside of the docker container, to a directory on the actual host.

It's possible (I think; I haven't tried it yet) to just disable the local cache. But now that I think about it further, I think it's a good idea, and I'd like to keep it.

But having 3 different local build caches on the same host, when they in the long term contain largely the same contents, seems wasteful. And has now led (on two separate occasions) to filling up the local disk.

I will probably work around this by limiting the size of the local cache. Apparently you can no longer limit by size, but I can change the default [DirectoryBuildCache.removeUnusuedEntriesAfterDays](https://docs.gradle.org/current/dsl/org.gradle.caching.local.DirectoryBuildCache.html#org.gradle.caching.local.DirectoryBuildCache:removeUnusedEntriesAfterDays) to be something less than the default of 7. But it would be preferable to just have all of the agents share a cache.

But from this issue it sounds like that's not possible.

**Comment by mmeijeri** [comment]
@oehme Is the workaround described by bsideup (--net=host Docker option) safe? If not, would it be safe to mount only .gradle/caches/modules-2 in the container? In that case each container would still have its own .gradle directory, but the modules-2 subdirectory would be mapped from the host.

Another potential workaround: would it be possible to add a command-line message to fall back to using the local Maven cache? I know Gradle already checks the local Maven cache first, but with such an option it could be instructed to write to it as well.

**Comment by zhleonix** [comment]
.gradle/caches/modules-2 can not be shared too as gradle will put a lock there once using it. What we have done is to setup a pool of shared caches and allocated to container to accelerate the build.

**Comment by nniesen** [comment]
We tried a mount to ~/.gradle/caches/modules-2/files-2.1 for just the artifacts (no build or metadata caching which we didn't want anyway). That seemed to work without causing locking issues. Stable artifacts did not re-download but snapshots with a 1ms changing timeout downloaded every time as expected.

YMMV but in our case, even though everything was working as expected, it didn't improve our pipeline build times. We are now thinking we have some other issue with our agents file system latency. Unlike others, we're not too concerned about disk space since the Jenkins agent containers are destroyed after every build.

**Comment by simpss** [comment]
> YMMV but in our case, even though everything was working as expected, it didn't improve our pipeline build times. We are now thinking we have some other issue with our agents file system latency. Unlike others, we're not too concerned about disk space since the Jenkins agent containers are destroyed after every build.

Using a different CI system(gitlab-ci) I had similar results. The real cause behind slow builds was IO. We migrated our gitlab runners(jenkins agents?) to SSD's and made sure every runner VM was on a different host to not have IO spikes at the same time.

**Comment by roadSurfer** [comment]
We are also wanting to make use of "throwaway" Jenkins Build Agents that are simply Docker Containers, but still share the cached downloads between them to save bandwidth and time.
We will often run multiple builds in parallel, and having to download & store the exact same artefact for the different builds is wasteful.
Or we will have to use multiple volumes and then run de-duplication/syncing on a schedule; which strikes me as brittle and prone to failure (and potentially blocking builds whilst they run).

**Comment by roadSurfer** [comment]
> We tried a mount to ~/.gradle/caches/modules-2/files-2.1 for just the artifacts (no build or metadata caching which we didn't want anyway). That seemed to work without causing locking issues. Stable artifacts did not re-download and snapshots with a 1ms changing timeout downloaded every time as expected.

@nniesen , how did you achieve that? When I try I see the ownership of "caches/modules-2" etc flip from "gradle:gradle" to "root:root" and the builds then fail.

**Comment by nniesen** [comment]
@roadSurfer: Unfortunately, I no longer have access to the environment but I believe everything was running as a build agent user on a Jenkins Kubernetes/Azure build agent. DevOps set up the Jenkins build agent configuration so that the jenkins agent's docker image had access to an external volume.

Perhaps your "throwaway" Jenkins Build Agent docker image was not properly setup to execute builds as the 'gradle' user. If the docker file doesn't create a user and switch to that user it will image will execute using the root user.

In the Jenkins pipeline, before the Agent ran a Gradle command that downloaded artifacts, I did something like the following:
```
# In the Jenkins pipeline before the Agent runs a Gradle command that downloads artifacts
# Note: /external-files-2.1 is the volume mount in the Agent that points to external location.

# In the Agents gradle user home, create the modules-2 directory.
mkdir -p ~/.gradle/caches/modules-2

# In the Agent, create the gradle user home files-2.1 link to the external directory.
ln -s /external-files-2.1 ~/.gradle/caches/modules-2/files-2.1
```

Now when the pipeline runs the `./gradlew build`, Gradle builds the metadata (in the Agents gradle home directory (`~/.gradle/`)) for all the existing artifacts in `/external-files-2.1` and also downloads any missing artifacts.

Gradle is usually very fast at rebuilding the metadata in the users and projects `.gradle` directory so I assume we were having latency issues accessing and computing file hashes for the artifacts on the external volume.

Note: You can kind of play around with it locally by setting the `GRADLE_USER_HOME` environment variable to a temporary location. Then you can arbitrarily delete files from that directory or the projects `.gradle` directory to see how Gradle gracefully rebuilds any missing information.

**Comment by roadSurfer** [comment]
Thanks @nniesen , I was wondering if the symlink trick would work. I placed the whole lot into a script to make my life easier:
```
#!/bin/bash

: ${GRADLE_USER_HOME='/home/gradle'}
GRADLE_EXTERNAL_CACHE=/gradle_artefacts
GRADLE_LINK_TARGET_PATH=${GRADLE_USER_HOME}/.gradle/caches/modules-2
GRADLE_LINK_TARGET_NAME=files-2.1

if [[ "${USE_CI}" == "Yes" ]]; then
	if [[ -L ${GRADLE_LINK_TARGET_PATH}/${GRADLE_LINK_TARGET_NAME} ]]; then
		echo "CI cache already enabled."
	else
		mkdir -p ${GRADLE_LINK_TARGET_PATH}
		ln -s ${GRADLE_EXTERNAL_CACHE} ${GRADLE_LINK_TARGET_PATH}/${GRADLE_LINK_TARGET_NAME}
	fi
else
	echo "CI not enabled. 'USE_CI' was: '${USE_CI}'"
fi
```
"USE_CI" should be set to "Yes" either in the image itself or via Container config.
The location "/gradle_artefacts" should be defined as a volume in the Image.

Only outstanding issue I can see with this approach is having to remember to run the script at the start of every pipeline. Still, it's an improvement.

**Comment by EugenMayer** [comment]
I created this https://discuss.gradle.org/t/sharing-managing-global-project-between-docker-containers-hosts-with-gradle5/31578 issue in the forum to maybe start a discussion about the use cases.  It also covers the topic of the project cache, which can be in your way too (in development rather then CI). 

I really wonder what the "reason" is behind https://github.com/gradle/gradle/issues/851#issuecomment-368226538 and also how this topic related to the "remote cache" https://docs.gradle.org/current/userguide/build_cache.html#sec:build_cache_configure_remote

I would love to have some exchange in the forum to not overload this issue too much, maybe if needed extract the core / moving forward into this or a different issue.

I think from a personal POV, this very much topic is harming the gradle project far more then anticipated - i already have seen a backoff from gradle after a successful migration from `maven` just because of that very reasons - they just went back to maven. Sharing `~/.m2` is just dead simple and those kind of cached folders are nodays very well supported by the CI solutions - where you just define a volume and give it a name, and it gets mounted on all your build docker containers.

It is practical on maven, npm, composer, and dep(go) and AFAIR also with go-modules from 1.17 - sounds like the demand is high and the reason understandable.

for @oehme i would really invite you to join the discussion there too. You suggestion with an nexus-proxy is by far not and "drop in alternative" - we are running a nexus-proxy anyway, and this is not even near to what you get with mounting a folder. Not only the transport time ( dedicated server in DC ), but the time to write i/o regarding to that amount of files is just not something one would suggest as "ok that is a viable alternative". Maybe you could reconsider here?

**Comment by cebaa** [comment]
This is not only a problem for containers, it's also a problem when used with "normal" hosts. E.g. say there are two Linux machines. You ssh into one and run `gradle build`. This starts a daemon on that host. When you ssh into the other one, it will not be able to proceed due to:

```
> Timeout waiting to lock file hash cache (/home/<user>/.gradle/caches/<version>/fileHashes). It is currently in use by another Gradle instance.
```

Even more problematic is that if you have a lot of hosts, you don't readily know which host the daemon is running on to kill it.

I agree with @EugenMayer's [comment](https://github.com/gradle/gradle/issues/851#issuecomment-487404773), this is really killing gradle as a development option.

**Comment by evilmilo** [comment]
Ah simple workaround for jenkins totally missed in earlier comments!

@redeamer commented on Mar 2, 2018
> In regard of this 'issue' in a CI context:
> For our setup I think I introduced a good 'workaround', but it depends on your number of executors per build node and the size of a singe cache: I map a container volume to keep the gradle caches and set the `GRADLE_USER_HOME` to `<cache_volume_path>/${env.EXECUTOR_NUMBER}` (on Jenkins-CI, do not know for other CIs). So I avoid any parallelism issues and have still the cache around for reuse and the cache duplication is justifiable/feasible (for us at least).

Can't believe missed this, such an obviously simple workaround! Issue only happens in Jenkins when parallel builds run and access the same cache.  So having one cache per executor means a cache can never be accessed by multiple containers at the same time.  We only have 2 executors, so only one extra cache no issue for us at all.

We are using jenkins pipeline, so the simple change we made to all our scripts was simply to change docker args from:
`args '-v gradleCache:/home/gradle/.gradle'`

to:
` args '-v gradleCache' + env.EXECUTOR_NUMBER + ':/home/gradle/.gradle'`

Thanks @redeamer!

**Comment by martinda** [comment]
> Even more problematic is that if you have a lot of hosts, you don't readily know which host the daemon is running on to kill it.

It would make sense for Gradle to store the host name and report it when the cache lock cannot be obtained. At least we would not have to hunt down which host holds the cache.

**Comment by yelhouti** [comment]
This clearly a much needed future, any ETA by any chance?

**Comment by EugenMayer** [comment]
I would expect there is rather a small interest in ever getting work done here because of Gradle EE and the build cache server. I think there are commercial interest steering away from this issue - which are find since somehow Gradle developers have to pay the bills too.

I must say, while there are not many othe reasons or none, this is the single most crushing one i will/would consider going back to maven

**Comment by matejdro** [comment]
Build cache server and Gradle EE does not have much to do with this though. This is mostly for dependency artifact cache while build cache server only caches gradle task steps.

**Comment by melix** [comment]
There is interest in fixing this and work is planned for 6.1, in whatever form it will take.

**Comment by omares** [comment]
Docker's [experimental `RUN --mount=type=cache` feature](https://github.com/moby/buildkit/blob/master/frontend/dockerfile/docs/experimental.md#run---mounttypecache) provides three `sharing` modes. The `private` or `locked` sound like they could solve the described problem:

> private creates a new mount if there are multiple writers. locked pauses the second writer until the first one releases the mount.

Did someone test these options in combination with a shared gradle cache?

**Comment by hnykda** [comment]
Coming from a different (python) world, I am somewhat shocked this is an issue. Gitlab CI, drone, concourse, github Actions... some of them being container-first and running e.g. a test, linter, checks and so on at once is a pretty neat way how to speed up builds. But because of this, it's impossible as individual (parallel) steps randomly fails. Not sharing the cache, on the other hand, takes enormous amount compare to the task (e.g. linting takes 10 seconds in our midsize project, downloading all deps on multi CPU AWS machine takes a minute at least)

**Comment by ljacomet** [comment]
FYI #1338, which somewhat relates to this issue, has been updated.

And there are plans to make the situation better overal:
(copied from #1338)

> Further ideas / improvements are under discussion:
>
> * Allow to add a read-only dependency cache layer. The goal would be to enable sharing that one between containers, while additional files not cached would still be downloaded inside the container. In addition to the network traffic reduction, this should also allow to reduce image sizes since the cache would not have to be embedded in each.
> * Provide some tooling around seeding, copying and setting up these different caches in a container environment.

**Comment by rgoomar** [comment]
Is this fixed with the changes in 6.1 or is there something else that will be released in 6.2 to resolve it?

**Comment by ljacomet** [comment]
Gradle 6.1 will support copying the dependency cache without constraint. Some form of sharing will be made available with Gradle 6.2. Expect an update here when it makes it into a nightly for early feedback.

**Comment by gluehbirnenkopf** [comment]
Just another workaround from my side: if you are using a docker image for building, just include the gradle cache in there, it will be unique per build. Of course regular updates need to happen

**Comment by rgoomar** [comment]
@gluehbirnenkopf That's exactly what we've done so far. A shared solution would be ideal to keep image sizes small and we have a bunch of projects that use similar dependencies.

**Comment by gluehbirnenkopf** [comment]
> @gluehbirnenkopf That's exactly what we've done so far. A shared solution would be ideal to keep image sizes small and we have a bunch of projects that use similar dependencies.

@rgoomar which folder exactly do you keep in your image? Complete .gradle/ folder ?

**Comment by melix** [comment]
Hi folks,

Gradle 6.2-rc-1 is out with an option to have a shareable, read-only dependency cache: https://docs.gradle.org/6.2-rc-1/userguide/dependency_resolution.html#sub:ephemeral-ci-cache

Please give it a try, we would appreciate any feedback on this before the final, thanks a lot!

**Comment by martinda** [comment]
Does gradle check whether the RO cache is effectively read-only, or does it trust the user on this?

> `$GRADLE_RO_DEP_CACHE`
   |-- modules-2 : the read-only dependency cache, should be mounted with read-only privileges

**Comment by melix** [comment]
It trusts the user that it's read-only (the docs mention that). Gradle _shouldn't_ write anything in there though, unless, of course, the directory you use is actually a write cache of another Gradle instance.

**Comment by matejdro** [comment]
Are there any strategies on how would re-seeding look like?

For example let's say a team updates dependency in one of their projects. I assume they would have to run some special build that would re-seed the cache and then copy it around?

**Comment by melix** [comment]
It's up to the team to decide. Some would have a build which re-seeds every day, others on every dependency change (maybe overkill), some on demand, ...

**Comment by pbarnoux** [comment]
Thanks for the feature.

Is it safe to have snapshots in the read only cache?
Does Gradle still honor changing resolution mechanisms? If a changing is too old both in the RW and RO dependencies caches, will it consult if newest builds are available on remote repositories?

**Comment by danielsuter** [comment]
How does Maven do it? I do not have permissions to move files into the read only cache (which is not unusual in large organizations). So my only option is too disable caching alltogether by using a specific directory for each build making builds performaing 2x slower than Maven builds.

**Comment by melix** [comment]
Maven doesn't do this. It's at your own risk, basically: you can have concurrent writes.

**Comment by danielsuter** [comment]
Would be nice if Gradle supported the same as it works perfectly well for maven. Gradle Remote Cache is much slower than disk cache. Keeping an additional read only cache adds quite some complexity.

**Comment by jiridanek** [comment]
Maven doesn't support concurrent writes, see [MNG-2802 Concurrent-safe access to local Maven repository](https://issues.apache.org/jira/browse/MNG-2802), open enhancement since Jan/07. In my experience, sharing `.m2` between concurrent builds always worked for me, except when `.m2` was on NFS. Maybe I was unusually lucky.

**Comment by melix** [comment]
> Would be nice if Gradle supported the same as it works perfectly well for maven

As stated here, it's not true. We have horror stories about concurrent writes in `.m2`, especially on CI. We _won't_ provide a way to do "unlocked writes" in Gradle, because it's inherently unsafe. That's precisely the reason for the feature we added the read-only dependency cache and the ability to relocate dependency caches.

Note that it has _nothing_ to do with the build cache. Saying "Gradle Remote Cache is much slower than disk cache" is irrelevant to this conversation, because the build cache is for _tasks execution_, not dependencies.

I just don't understand that either:

>  So my only option is too disable caching alltogether

as I said, tasks caching has nothing in common to the conversation here.

**Comment by martinda** [comment]
It would help me if I understood where gradle puts different types of files, and whether those locations support concurrent access. I feel that users like me have some understanding of that, but perhaps it is incomplete. It would help me if I could find the information on a single page explaining the caches, the repositories and the other places gradle accesses for read or write, for example: `.m2`, `.ivy2`, `~/.gradle`,  `.gradle`, the build folder, `/tmp`, etc.

What seems to be harder to understand is whether those are designed for concurrent access or not, how and when to clean them, and the conditions under which they are locked (lock acquired and released).

I often work on multiple projects at the same time on multiple hosts simultaneously all using the same NFS mounted /home or NFS mounted build folder. It would be helpful to understand where the caches and repositories should be located (on a disk local to an execution host?, on an NFS mounted on to multiple execution hosts?). Examples would also help. For example, should I share the `.m2` folder with builds running in maven and gradle at the same time accessing the same `.m2` folder?

**Comment by brunojcm** [comment]
It seems maven is not dead after all, [Concurrent-safe access to local Maven repository](https://issues.apache.org/jira/browse/MNG-2802?focusedCommentId=17178879&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-17178879) is now fixed :)

**Comment by hongpuwu** [comment]
> It's up to the team to decide. Some would have a build which re-seeds every day, others on every dependency change (maybe overkill), some on demand, ...

when re-seed，what files should I copy to the GRADLE_RO_DEP_CACHE？there maybe a lost of cache dir to write to GRADLE_RO_DEP_CACHE

**Comment by eryajf** [comment]
Now, in 2023, may I ask if there is a relatively perfect solution to this problem?

I compiled the project in Jenkins as follows：

```groovy
def buildJavaJar(BUILD_BASE_IMAGE, BUILD_ARGS, BUILD_COMMAND){
    try {
        docker.image("eryajf/gradle").inside("-v /root/.gradle:/root/.gradle") {
            sh "gradle"
        }
    }catch(exc) {
        env.FAILURE_MSG = "build failed"
        throw(exc)
    }
}
```

This problem is bound to occur when multiple projects are built at the same time.

**Comment by travisdowns** [comment]
One thing that I think is unclear in the discussion is that gradle _already_ supports concurrent writes into the modules-2 module cache: that's what the .lock file is there for an why the error is about an inability to lock that file. So docker containers which share a --mount=type=cache should be supported. It _only_ doesn't work because in the container scenario gradle instances try to communicate over the localhost network to talk to other gradle instances and if they don't see any, they just hold the lock for a entire build instead of releasing it in a fine-grained manner, or something like that.

**Comment by cobexer** [comment]
A summary of the current state so it can be better linked from the README.md:

Sharing of the Gradle User Home between containers / with the host is supported thus this issue is closed fixed.
However concurrently (from running builds) using the cache does not lead to optimal performance as Gradle coordinates fine grained locking of the caches using local network connections which may not be available depending on how containers are launched.

There is a feature request over at https://github.com/gradle/gradle/issues/31394 for moving the communication over to unix domain sockets which (AFAIK) could help this use case as well.

**Comment by adamscybot** [comment]
Thanks @cobexer. I had trouble finding what I would need to configure to make this work across multiple containers. For the sake of anyone else coming from the readme I'll point out [this useful comment](https://github.com/gradle/gradle/issues/8375#issuecomment-3083654179) which states it needs access to the host loopback interface.

However, I am a bit confused on the official position about sharing the user home dir across containers because of [this comment](https://github.com/gradle/gradle/issues/8375#issuecomment-3083466984). Possibly a change to the docs would be good to clean up the confusion?

Thank you for your support.

**Comment by diktamxx** [comment]
how this issue is resolved in Gradle 9 nowadays?
My approach was to share the ${GRADLE_USER_HOME}/caches/modules-2/files-2.1 file across all layers of the Dockerfile (I did not use RUN --mount because this seems not to be a standard feature). However, currently I am encountering the same problem as [https://github.com/gradle/gradle/issues/1338#issuecomment-1077740068](https://github.com/gradle/gradle/issues/1338#issuecomment-1077740068). 

This is a snippet of the relevant definition:
```
FROM gradle:9.2.1-jdk25-alpine AS dependencies
WORKDIR /app
COPY build.gradle settings.gradle ./
RUN gradle dependencies --no-daemon

FROM gradle:9.2.1-jdk25-alpine AS distribution
ENV GRADLE_USER_HOME=/root/.gradle
COPY --from=dependencies ${GRADLE_USER_HOME}/caches/modules-2/files-2.1 ${GRADLE_USER_HOME}/caches/modules-2/files-2.1
```
error message:
```
19.60 FAILURE: Build failed with an exception.
19.60 
19.60 * Where:
19.60 Build file '/app/build.gradle' line: 14
19.60 
19.60 * What went wrong:
19.60 Plugin [id: 'co.uzzu.dotenv.gradle', version: '4.0.0'] was not found in any of the following sources:
19.60 
19.60 - Gradle Core Plugins (plugin is not in 'org.gradle' namespace)
19.60 - Included Builds (No included builds contain this plugin)
19.60 - Plugin Repositories (could not resolve plugin artifact 'co.uzzu.dotenv.gradle:co.uzzu.dotenv.gradle.gradle.plugin:4.0.0')
19.60   Searched in the following repositories:
19.60     Gradle Central Plugin Repository

```

**Comment by travisdowns** [comment]
@diktamxx my solution was to use a separate cache mount for each docker layer that might execute in parallel, e.g.:

```
RUN --mount=type=cache,target=/root/.gradle/caches/modules-2,id=some-layer
```

Change `id=some-layer` to a different ID in each layer that might conflict. I means you end up with N caches instead of 1, but this, for us, was still much faster than no cache at all (on rebuilds, since the gradle cache is already populated for each layer), at the cost of additional disk space.

**Comment by sesamzoo** [comment]
Thanks for sharing, @travisdowns.

**Comment by diktamxx** [comment]
> [@diktamxx](https://github.com/diktamxx) my solution was to use a separate cache mount for each docker layer that might execute in parallel, e.g.:
> 
> ```
> RUN --mount=type=cache,target=/root/.gradle/caches/modules-2,id=some-layer
> ```
> 
> Change `id=some-layer` to a different ID in each layer that might conflict. I means you end up with N caches instead of 1, but this, for us, was still much faster than no cache at all (on rebuilds, since the gradle cache is already populated for each layer), at the cost of additional disk space.

You're right. I've tried RUN --mount before, and it really did improve the build speed. But the idea is to avoid duplicate builds. Because build.gradle and settings.gradle don't change very often, you don't need to re-download the system's dependencies every time (which is relatively time-consuming). But there seems to be no good solution. I found that copying caches/caches/modules-2 did not allow me to reuse previously downloaded dependencies.

**Comment by travisdowns** [comment]
> I found that copying caches/caches/modules-2 did not allow me to reuse previously downloaded dependencies.

Not sure what's going on for you, but it works great for me as described above.

Once I download the dependencies once, they are not downloaded again as long as the cache lives.

Maybe your cache is getting garbage collected: with default settings this can happen basically right after your build if the image is more than about 20GB or so.

You can do an `find /root/.gradle/caches` in your Dockerfile, in a step with the cache mounted, to see if it is actually populated or has been GC'd by docker.

**Comment by travisdowns** [comment]
> But the idea is to avoid duplicate builds. 

Oh do you mean avoid duplicates between different layers within a single docker build?

**Comment by diktamxx** [comment]
> > But the idea is to avoid duplicate builds.
> 
> Oh do you mean avoid duplicates between different layers within a single docker build?

Yes. I think it's possible to separate the "download dependencies" operation into its own layer. Because (build.gradle, settings.gradle) rarely changes, once it is reused, it avoids the need to download dependencies every time an image is built.

**Comment by travisdowns** [comment]
> Yes. I think it's possible to separate the "download dependencies" operation into its own layer. Because (build.gradle, settings.gradle) rarely changes, once it is reused, it avoids the need to download dependencies every time an image is built.

Yes, that should work fine though? No need to for --mount or cache stuff. Put your download step as a parent layer and it should work?

**Comment by diktamxx** [comment]
> > Yes. I think it's possible to separate the "download dependencies" operation into its own layer. Because (build.gradle, settings.gradle) rarely changes, once it is reused, it avoids the need to download dependencies every time an image is built.
> 
> Yes, that should work fine though? No need to for --mount or cache stuff. Put your download step as a parent layer and it should work?

It doesn't work. It seems that the Gradle cache cannot be copied or reused at will. [https://github.com/gradle/gradle/issues/1338](https://github.com/gradle/gradle/issues/1338)

**Comment by travisdowns** [comment]
> It doesn't work. It seems that the Gradle cache cannot be copied or reused at will. [#1338](https://github.com/gradle/gradle/issues/1338)

I don't think that's related. If you have 1 RUN command that populates the cache, then another RUN command that uses it (e.g., for a compile) it, it will work fine: there is no relocation or copying, gradle can't even know about the layers. From gradle PoV it's just a normal cache re-use between serial commands.

_This specific_ issue is about multiple containers or layers in a single build trying to use the same cache _concurrently_ and running into issues with locking, and that the locking coordination approach (which does let grade processes actively coordinate access to the cache) doesn't work when container are involved (because the gradle processes are effectively invisible to each other).

Your issue sounds different and IMO it should work fine. Maybe consider a simple reproducer and a new issue? It's more likely to work than this thread on a closed issue in any case.
